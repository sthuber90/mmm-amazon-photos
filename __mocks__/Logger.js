global.Log = {
  debug: jest.fn(),
  log: jest.fn(),
  info: jest.fn(),
  warn: jest.fn(),
  error: jest.fn(),
}
