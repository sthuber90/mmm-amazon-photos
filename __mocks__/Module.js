global.Module = {
  definitions: {},
  create(name) {
    return this.definitions[name]
  },
  register(name, overrides) {
    const base = {
      defaults: {},

      showHideTimer: null,

      lockStrings: [],

      _nunjucksEnvironment: { addFilter: jest.fn() },

      init() {
        Log.log(this.defaults)
      },

      start() {
        Log.info('Starting module: ' + this.name)
      },

      getScripts() {
        return []
      },

      getStyles() {
        return []
      },

      getTranslations() {
        return false
      },

      getDom: jest.fn(),

      getHeader() {
        return this.data.header
      },

      getTemplate() {
        return (
          '<div class="normal">' +
          this.name +
          '</div><div class="small dimmed">' +
          this.identifier +
          '</div>'
        )
      },

      getTemplateData() {
        return {}
      },

      notificationReceived(notification, payload, sender) {
        if (sender) {
          Log.log(
            this.name +
              ' received a module notification: ' +
              notification +
              ' from sender: ' +
              sender.name
          )
        } else {
          Log.log(
            this.name + ' received a system notification: ' + notification
          )
        }
      },

      nunjucksEnvironment() {
        return this._nunjucksEnvironment
      },

      socketNotificationReceived(notification, payload) {
        Log.log(
          this.name +
            ' received a socket notification: ' +
            notification +
            ' - Payload: ' +
            payload
        )
      },

      suspend() {
        Log.log(this.name + ' is suspended.')
      },

      resume() {
        Log.log(this.name + ' is resumed.')
      },

      setData(data) {
        this.data = data
        this.name = data.name
        this.identifier = data.identifier
        this.hidden = false

        this.setConfig(data.config, data.configDeepMerge)
      },

      setConfig(config) {
        this.config = Object.assign({}, this.defaults, config)
      },

      translate: jest.fn(),

      updateDom: jest.fn(),

      sendNotification: jest.fn(),

      sendSocketNotification: jest.fn(),

      hide: jest.fn((speed, callback) => {
        if (typeof callback === 'object') {
          callback = jest.fn()
        }

        callback = callback || jest.fn()

        global.Module.definitions[name].hidden = true
        global.Module.definitions[name].suspend()
        callback()
      }),

      show: jest.fn((speed, callback) => {
        if (typeof callback === 'object') {
          callback = jest.fn()
        }

        callback = callback || jest.fn()

        global.Module.definitions[name].hidden = false
        global.Module.definitions[name].resume()
        callback()
      }),
      file: (filename) => {
        return filename
      },
    }

    this.definitions[name] = { ...base, ...overrides }
  },
}
