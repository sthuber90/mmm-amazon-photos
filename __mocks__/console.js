const originalConsole = { ...global.console }

function mockConsole() {
  global.console.log = jest.fn()
  global.console.warn = jest.fn()
  global.console.info = jest.fn()
  global.console.error = jest.fn()
}

function restoreConsole() {
  global.console = originalConsole
}

module.exports = { mockConsole, restoreConsole }
