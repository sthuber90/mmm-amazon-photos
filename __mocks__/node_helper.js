module.exports = {
  create(overrides) {
    const base = {
      init() {
        console.log('Initializing new module helper ...')
      },

      loaded(callback) {
        console.log(`Module helper loaded: ${this.name}`)
        callback()
      },

      start() {
        console.log(`Starting module helper: ${this.name}`)
      },

      stop() {
        console.log(`Stopping module helper: ${this.name}`)
      },

      socketNotificationReceived(notification, payload) {
        console.log(
          `${this.name} received a socket notification: ${notification} Payload: ${payload}`
        )
      },

      setName(name) {
        this.name = name
      },

      setPath(path) {
        this.path = path
      },

      sendSocketNotification: jest.fn(),

      setExpressApp: jest.fn(),

      setSocketIO: jest.fn(),
    }

    return { ...base, ...overrides }
  },
}
