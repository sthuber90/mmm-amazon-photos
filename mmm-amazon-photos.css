.mmm-amazon-photos .images,
.mmm-amazon-photos .image,
.mmm-amazon-photos .gradient {
  position: absolute;
  top: 0;
  left: 0;
  height: 100%;
  width: 100%;
}
.mmm-amazon-photos .image {
  background-repeat: no-repeat;
}

.mmm-amazon-photos .info {
  bottom: 2rem;
  right: 2rem;
  display: none;
  padding: 15px;
  font-size: var(--font-size);
  line-height: 25px;
  position: absolute;
  grid-template-columns: auto auto;
  gap: 0px 10px;
}

.mmm-amazon-photos .info .icon {
  display: inline-block;
  line-height: 25px;
  height: var(--font-size);
  width: var(--font-size);
  vertical-align: baseline;
}
