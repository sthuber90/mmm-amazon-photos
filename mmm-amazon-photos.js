/* global Module */

/* mmm-amazon-photos.js
 *
 * Magic Mirror
 * Module: mmm-amazon-photos
 *
 * Magic Mirror By Michael Teeuw http://michaelteeuw.nl
 * MIT Licensed.
 *
 * Module mmm-amazon-photos by Stephan Huber
 * MIT Licensed.
 */

Module.register('mmm-amazon-photos', {
  // Default module config.
  defaults: {
    // if set to true adds a click handler to the gradient to toggle other module visibility.
    // works only together with the gradient
    focus: true,
    // an array of strings, each is a path to a directory with images
    imageUrls: ['https://www.amazon.com/clouddrive/share/placeholderShareId'],
    // the speed at which to switch between images, in milliseconds
    slideshowSpeed: 60 * 60 * 1000,
    // the sizing of the background image
    // cover: Resize the background image to cover the entire container, even if it has to stretch the image or cut a little bit off one of the edges
    // contain: Resize the background image to make sure the image is fully visible
    backgroundSize: 'cover', // cover or contain
    // if backgroundSize contain, determine where to zoom the picture. Towards top, center or bottom
    backgroundPosition: 'center', // Most useful options: "top" or "center" or "bottom"
    // the gradient to make the text more visible
    gradient: [
      'rgba(0, 0, 0, 0.75) 0%',
      'rgba(0, 0, 0, 0) 40%',
      'rgba(0, 0, 0, 0) 80%',
      'rgba(0, 0, 0, 0.75) 100%',
    ],
    // the direction the gradient goes, vertical, horizontal or both for a vignette effect
    gradientDirection: 'vertical',
  },

  // load function
  start: function () {
    // add identifier to the config
    this.config.identifier = this.identifier
  },

  getScripts: function () {
    return []
  },

  getStyles: function () {
    // the css contains the make grayscale code
    return [this.file('mmm-amazon-photos.css')]
  },

  // generic notification handler
  notificationReceived: function (notification, payload, sender) {
    if (sender) {
      if (notification === 'AMAZONPHOTOS_IMAGE_UPDATE') {
        Log.log('mmm-amazon-photos: Changing Background')
        this.suspend()
        this.updateImage()
        this.resume()
      } else if (notification === 'AMAZONPHOTOS_PLAY') {
        // Change to next image and start timer.
        this.updateImage()
        this.resume()
      } else if (notification === 'AMAZONPHOTOS_PAUSE') {
        // Stop timer.
        this.suspend()
      } else if (notification === 'AMAZONPHOTOS_URL') {
        if (payload && payload.url) {
          // Stop timer.
          if (payload.resume) {
            if (this.timer) {
              // Restart timer only if timer was already running
              this.resume()
            }
          } else {
            this.suspend()
          }
          this.updateImage(payload.url)
          // no urls sent, see if there is saved data.
        } else if (this.savedImage) {
          this.image = this.savedImage
          this.savedImage = null
          this.updateImage()
          if (this.timer) {
            // Restart timer only if timer was already running
            this.resume()
          }
        }
      }
    }
  },

  // the socket handler
  socketNotificationReceived: function (notification, payload) {
    // if an update was received
    if (notification === 'AMAZONPHOTOS_FILELIST') {
      // check this is for this module based on the id
      if (payload.identifier === this.identifier) {
        // set the image list
        if (this.savedImage) {
          this.savedImage = payload.image
        } else {
          this.image = payload.image
          // if image list actually contains images
          // set loaded flag to true and update dom
          if (typeof this.image !== 'undefined' && this.image !== '') {
            this.updateImage() //Added to show the image at least once, but do not change it within this.resume()
            this.resume()
          }
        }
      }
    }
  },

  // Override dom generator.
  getDom: function () {
    const wrapper = document.createElement('div')
    this.imagesDiv = document.createElement('div')
    this.imagesDiv.className = 'images'
    wrapper.appendChild(this.imagesDiv)

    if (
      this.config.gradientDirection === 'vertical' ||
      this.config.gradientDirection === 'both'
    ) {
      const gradientBottom = this.createGradientDiv(
        'bottom',
        this.config.gradient
      )
      wrapper.appendChild(gradientBottom)
    }

    if (
      this.config.gradientDirection === 'horizontal' ||
      this.config.gradientDirection === 'both'
    ) {
      const gradientRight = this.createGradientDiv(
        'right',
        this.config.gradient
      )
      wrapper.appendChild(gradientRight)
    }

    if (this.config.imageUrls.length === 0) {
      Log.error('mmm-amazon-photos: Missing required parameter imageUrls.')
    } else {
      // create an empty image list
      this.image = ''
      this.retrieveImage()
    }

    return wrapper
  },

  createGradientDiv: function (direction, gradient) {
    const div = document.createElement('div')
    div.style.backgroundImage =
      'linear-gradient( to ' + direction + ', ' + gradient.join() + ')'
    div.className = 'gradient'
    this.config.focus
      ? div.addEventListener('click', function () {
          Array.from(document.getElementsByClassName('container')).forEach(
            function (elem) {
              if (
                !elem.innerHTML.includes('mmm-amazon-photos') &&
                elem.innerHTML !== ''
              ) {
                if (elem.style.display === 'none') {
                  elem.style.display = 'block'
                } else {
                  elem.style.display = 'none'
                }
              }
            }
          )
        })
      : undefined
    return div
  },

  createDiv: function () {
    const div = document.createElement('div')
    div.style.backgroundSize = this.config.backgroundSize
    div.style.backgroundPosition = this.config.backgroundPosition
    div.className = 'image'
    return div
  },

  updateImage: function (imageToDisplay = null) {
    if (!imageToDisplay) {
      if (!this.image || this.image === '') {
        return
      }
    }

    const image = new Image()

    if (imageToDisplay) {
      image.src = encodeURI(imageToDisplay)
    } else {
      image.src = encodeURI(this.image)
    }

    const that = this
    image.onload = function () {
      // check if there are more than 2 elements and remove the first one
      if (that.imagesDiv.childNodes.length > 1) {
        that.imagesDiv.removeChild(that.imagesDiv.childNodes[0])
      }
      if (that.imagesDiv.childNodes.length > 0) {
        that.imagesDiv.childNodes[0].style.opacity = '0'
      }

      const imageDiv = that.createDiv()
      imageDiv.style.backgroundImage = `url("${image.src}")`
      that.imagesDiv.appendChild(imageDiv)
    }
  },

  suspend: function () {
    if (this.timer) {
      clearInterval(this.timer)
      this.timer = null
    }
  },

  resume: function () {
    this.suspend()
    const self = this
    this.timer = setInterval(function () {
      Log.info(`${self.name}: updating from resume`)
      self.retrieveImage()
    }, self.config.slideshowSpeed)
  },

  retrieveImage: function () {
    this.suspend()
    Log.info('mmm-amazon-photos: Getting Image')
    // ask helper function to get the image list
    this.sendSocketNotification('AMAZONPHOTOS_REGISTER_CONFIG', this.config)
  },
})
