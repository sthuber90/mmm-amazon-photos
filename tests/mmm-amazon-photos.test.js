describe('mmm-amazon-photos', () => {
  beforeAll(() => {
    require('../__mocks__/Logger')
    require('../__mocks__/Module')
  })

  const name = 'mmm-amazon-photos'

  let MMMAmazonPhotos

  beforeEach(() => {
    jest.resetModules()
    require('../mmm-amazon-photos')

    MMMAmazonPhotos = global.Module.create(name)
    MMMAmazonPhotos.setData({ name, identifier: `Module_1_${name}` })
  })

  it('should have correct defaults', () => {
    expect(MMMAmazonPhotos.defaults).toMatchSnapshot()
  })

  it('should start', () => {
    MMMAmazonPhotos.start()
    expect(MMMAmazonPhotos.config.identifier).toEqual(
      'Module_1_mmm-amazon-photos'
    )
  })

  it('should get styles', () => {
    expect(MMMAmazonPhotos.getStyles()).toEqual(['mmm-amazon-photos.css'])
  })

  it('should get scripts', () => {
    expect(MMMAmazonPhotos.getScripts()).toEqual([])
  })

  it('should update image on notification received', () => {
    MMMAmazonPhotos.config.slideshowSpeed = 0
    MMMAmazonPhotos.notificationReceived(
      'AMAZONPHOTOS_IMAGE_UPDATE',
      '',
      'sender'
    )
    expect(global.Log.log).toHaveBeenNthCalledWith(
      1,
      'mmm-amazon-photos: Changing Background'
    )
  })

  it('should get DOM', () => {
    expect(MMMAmazonPhotos.getDom()).toMatchSnapshot()
  })

  it('should get DOM', () => {
    expect(MMMAmazonPhotos.getDom()).toMatchSnapshot()
  })

  it('should create gradient div', () => {
    const getElementsByClassNameSpy = jest.spyOn(
      document,
      'getElementsByClassName'
    )
    const innerDiv = document.createElement('div')
    getElementsByClassNameSpy.mockReturnValue([innerDiv])

    const div = MMMAmazonPhotos.createGradientDiv('bottom', [])
    expect(div).toMatchSnapshot()

    div.click()
    expect(innerDiv.style.display).toEqual('')

    innerDiv.innerHTML = 'foobar'
    div.click()
    expect(innerDiv.style.display).toEqual('none')

    div.click()
    expect(innerDiv.style.display).toEqual('block')
  })

  it('should create gradient div without click listener', () => {
    MMMAmazonPhotos.config.focus = false
    const getElementsByClassNameSpy = jest.spyOn(
      document,
      'getElementsByClassName'
    )
    const innerDiv = document.createElement('div')
    getElementsByClassNameSpy.mockReturnValue([innerDiv])

    const div = MMMAmazonPhotos.createGradientDiv('bottom', [])
    expect(div).toMatchSnapshot()

    div.click()
    expect(innerDiv.style.display).toEqual('')
  })

  it('should create div', () => {
    expect(MMMAmazonPhotos.createDiv()).toMatchSnapshot()
  })

  it('should update image and return undefined', () => {
    MMMAmazonPhotos.image = null
    expect(MMMAmazonPhotos.updateImage('a')).toBeUndefined()

    MMMAmazonPhotos.image = ''
    expect(MMMAmazonPhotos.updateImage('')).toBeUndefined()
  })

  it('should not supsend nor clear interval', () => {
    const clearIntervalSpy = jest.spyOn(global.window, 'clearInterval')
    MMMAmazonPhotos.suspend()
    expect(MMMAmazonPhotos.timer).toBeUndefined()
    expect(clearIntervalSpy).not.toHaveBeenCalled()
  })

  it('should supsend and clear interval', () => {
    const clearIntervalSpy = jest.spyOn(global.window, 'clearInterval')
    MMMAmazonPhotos.timer = 123
    MMMAmazonPhotos.suspend()
    expect(MMMAmazonPhotos.timer).toBeNull()
    expect(clearIntervalSpy).toHaveBeenCalled()
  })

  it('should resume', () => {
    MMMAmazonPhotos.config.slideshowSpeed = 0
    jest.useFakeTimers()
    MMMAmazonPhotos.resume()
    jest.advanceTimersByTime(1000)
    expect(global.Log.info).toHaveBeenNthCalledWith(
      1,
      'mmm-amazon-photos: updating from resume'
    )
  })

  it('should retrieve image', () => {
    MMMAmazonPhotos.retrieveImage()
    expect(global.Log.info).toHaveBeenNthCalledWith(
      1,
      'mmm-amazon-photos: Getting Image'
    )
    expect(MMMAmazonPhotos.sendSocketNotification).toHaveBeenNthCalledWith(
      1,
      'AMAZONPHOTOS_REGISTER_CONFIG',
      {
        backgroundPosition: 'center',
        backgroundSize: 'cover',
        focus: true,
        gradient: [
          'rgba(0, 0, 0, 0.75) 0%',
          'rgba(0, 0, 0, 0) 40%',
          'rgba(0, 0, 0, 0) 80%',
          'rgba(0, 0, 0, 0.75) 100%',
        ],
        gradientDirection: 'vertical',
        imageUrls: [
          'https://www.amazon.com/clouddrive/share/placeholderShareId',
        ],
        slideshowSpeed: 3600000,
      }
    )
  })
})
