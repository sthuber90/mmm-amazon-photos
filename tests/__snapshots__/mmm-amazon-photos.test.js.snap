// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mmm-amazon-photos should create div 1`] = `
<div
  class="image"
  style="background-size: cover; background-position: center;"
/>
`;

exports[`mmm-amazon-photos should create gradient div 1`] = `
<div
  class="gradient"
/>
`;

exports[`mmm-amazon-photos should create gradient div without click listener 1`] = `
<div
  class="gradient"
/>
`;

exports[`mmm-amazon-photos should get DOM 1`] = `
<div>
  <div
    class="images"
  />
  <div
    class="gradient"
  />
</div>
`;

exports[`mmm-amazon-photos should get DOM 2`] = `
<div>
  <div
    class="images"
  />
  <div
    class="gradient"
  />
</div>
`;

exports[`mmm-amazon-photos should have correct defaults 1`] = `
Object {
  "backgroundPosition": "center",
  "backgroundSize": "cover",
  "focus": true,
  "gradient": Array [
    "rgba(0, 0, 0, 0.75) 0%",
    "rgba(0, 0, 0, 0) 40%",
    "rgba(0, 0, 0, 0) 80%",
    "rgba(0, 0, 0, 0.75) 100%",
  ],
  "gradientDirection": "vertical",
  "imageUrls": Array [
    "https://www.amazon.com/clouddrive/share/placeholderShareId",
  ],
  "slideshowSpeed": 3600000,
}
`;
